imap/imap.go

```diff
@@ -150,203 +61,150 @@ func Validate(s *models.IMAP) error {
+	messages := make(chan *imap.Message)
 
-	var email Email
-	for _, msgData := range fCmd.Data {
-		msgFields := msgData.MessageInfo().Attrs
-
-		// make sure is a legit response before we attempt to parse it
-		// deal with unsolicited FETCH responses containing only flags
-		// I'm lookin' at YOU, Gmail!
-		// http://mailman13.u.washington.edu/pipermail/imap-protocol/2014-October/002355.html
-		// http://stackoverflow.com/questions/26262472/gmail-imap-is-sometimes-returning-bad-results-for-fetch
-		if _, ok := msgFields["RFC822.HEADER"]; !ok {
-			continue
+	go func() {
+		if err := imapClient.Fetch(seqset, items, messages); err != nil {
+			log.Error("Error fetching emails: ", err.Error()) // TODO: How to handle this, need to propogate error out
 		}
+	}()
 
-		email, err = NewEmail(msgFields)
-		if err != nil {
-			responses <- Response{Err: fmt.Errorf("unable to parse email: %s", err)}
-			return
+	// Step through each email
+	for msg := range messages {
+		// Extract raw message body. I can't find a better way to do this with the emersion library
+		var em *email.Email
+		var buf []byte
+		for _, value := range msg.Body {
+			buf = make([]byte, value.Len())
+			value.Read(buf)
+			break // There should only ever be one item in this map, but I'm not 100% sure
 		}
 
-		responses <- Response{Email: email}
-
-		if !markAsRead {
-			err = removeSeen(client, imap.AsNumber(msgFields["UID"]))
-			if err != nil {
-				responses <- Response{Err: fmt.Errorf("unable to remove seen flag: %s", err)}
-				return
-			}
-		}
+		//Remove CR characters, see https://github.com/jordan-wright/email/issues/106
+		tmp := string(buf)
+		re := regexp.MustCompile(`\r`)
+		tmp = re.ReplaceAllString(tmp, "")
+		buf = []byte(tmp)
 
-		if delete {
-			err = deleteEmail(client, imap.AsNumber(msgFields["UID"]))
-			if err != nil {
-				responses <- Response{Err: fmt.Errorf("unable to delete email: %s", err)}
-				return
-			}
+		rawBodyStream := bytes.NewReader(buf)
+		em, err = email.NewEmailFromReader(rawBodyStream) // Parse with @jordanwright's library
+		if err != nil {
+			return emails, err
 		}
-	}
-	return
-}
 
-func deleteEmail(client Client, UID uint32) error {
-	return alterEmail(client, UID, "\\DELETED", true)
-}
+		emtmp := Email{Email: em, SeqNum: msg.SeqNum} // Not sure why msg.Uid is always 0, so swapped to sequence numbers
+		emails = append(emails, emtmp)
 
-func removeSeen(client Client, UID uint32) error {
-	return alterEmail(client, UID, "\\SEEN", false)
+	}
+	return emails, nil
 }
 
-func alterEmail(client Client, UID uint32, flag string, plus bool) error {
-	flg := "-FLAGS"
-	if plus {
-		flg = "+FLAGS"
+// newClient will initiate a new IMAP connection with the given creds.
+func (mbox *Mailbox) newClient() (*client.Client, error) {
+	var imapClient *client.Client
+	var err error
+	if mbox.TLS {
+		imapClient, err = client.DialTLS(mbox.Host, new(tls.Config))
+		if err != nil {
```

Comment:
Small nit - we can move the `if err != nil` check outside the `if mbox.TLS` conditional to avoid repeating ourselves.

```go
if mbox.TLS {
    imapClient, err = client.DialTLS(mbox.Host, new(tls.Config))
} else {
    imapClient, err = client.Dial(mbox.Host)
}
if err != nil {
    return imapClient, err
}
```